python/cudf/cudf/core/column/datetime.py

```diff
@@ -630,6 +630,38 @@ def quantile(
             )
         return result.astype(self.dtype)
 
+    def find_and_replace(
+        self,
+        to_replace: ColumnBase,
+        replacement: ColumnBase,
+        all_nan: bool = False,
+    ) -> DatetimeColumn:
+        if not isinstance(to_replace, DatetimeColumn):
+            to_replace = cudf.core.column.as_column(to_replace)
+            if to_replace.can_cast_safely(self.dtype):
+                to_replace = to_replace.astype(self.dtype)
+        if not isinstance(replacement, DatetimeColumn):
+            replacement = cudf.core.column.as_column(replacement)
+            if replacement.can_cast_safely(self.dtype):
+                replacement = replacement.astype(self.dtype)
+        if isinstance(to_replace, DatetimeColumn):
+            to_replace = to_replace.as_numerical_column(
+                dtype=np.dtype("int64")
+            )
+        if isinstance(replacement, DatetimeColumn):
+            replacement = replacement.as_numerical_column(
+                dtype=np.dtype("int64")
+            )
```

Comment:
Can these two if statements not be else clauses of the first two conditionals because those might produce DatetimeColumns from the astype calls and we need to unconditionally convert to numerical here whether the original input was a DatetimeColumn or one was produced from astype?

Reply:
These can't be in the else clauses because of inputs like this: `DatetimeColumn.find_and_replace(to_replace=StringColumn, replacement=NumericalColumn)` if clauses above. 

If they were to be packed together, it has to be modified like:

```python
if not isinstance(to_replace, DatetimeColumn):
            to_replace = cudf.core.column.as_column(to_replace)
            if to_replace.can_cast_safely(self.dtype):
                to_replace = to_replace.astype(self.dtype).as_numerical_column(
                    dtype=np.dtype("int64")
                )
        else:
            to_replace = to_replace.as_numerical_column(
                dtype=np.dtype("int64")
            )
```

Reply:
Yeah right so you end up needing to call as_numerical_column anyway, that's what I was getting at. Cool, thanks.